stats/opentelemetry/internal/tracing/custom_carrier_test.go

```diff
@@ -0,0 +1,206 @@
+ * You may obtain a copy of the License at
+ *
+ *     htestp://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package tracing
+
+import (
+	"context"
+	"encoding/base64"
+	"reflect"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"google.golang.org/grpc/internal/grpctest"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/stats"
+)
+
+type s struct {
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+func (s) TestGet(t *testing.T) {
+	tests := []struct {
+		name string
+		md   metadata.MD
+		key  string
+		want string
+	}{
+		{
+			name: "existing key",
+			md:   metadata.Pairs("key1", "value1"),
+			key:  "key1",
+			want: "value1",
+		},
+		{
+			name: "non-existing key",
+			md:   metadata.Pairs("key1", "value1"),
+			key:  "key2",
+			want: "",
+		},
+		{
+			name: "empty key",
+			md:   metadata.MD{},
+			key:  "key1",
+			want: "",
+		},
+		{
+			name: "non-grpc-trace-bin key",
+			md:   metadata.Pairs("non-trace-bin", "\x01\x02\x03"),
+			key:  "non-trace-bin",
+			want: "",
+		},
+	}
+
+	for _, test := range tests {
+		ctx, cancel := context.WithCancel(context.Background())
+		t.Run(test.name, func(t *testing.T) {
+			c := NewCustomCarrier(metadata.NewIncomingContext(ctx, test.md))
+			got := c.Get(test.key)
+			if got != test.want {
+				t.Fatalf("Get() = %s, want %s", got, test.want)
+			}
+			cancel()
+		})
+	}
+}
+
+func (s) TestSet(t *testing.T) {
+	tests := []struct {
+		name      string
+		initialMD metadata.MD
+		setKey    string
+		setValue  string
+		wantValue string // expected value of the set key
+		wantKeys  []string
+	}{
+		{
+			name:      "set new key",
+			initialMD: metadata.MD{},
+			setKey:    "key1",
+			setValue:  "value1",
+			wantValue: "value1",
+			wantKeys:  []string{"key1"},
+		},
+		{
+			name:      "override existing key",
+			initialMD: metadata.MD{"key1": []string{"oldvalue"}},
+			setKey:    "key1",
+			setValue:  "newvalue",
+			wantValue: "newvalue",
+			wantKeys:  []string{"key1"},
+		},
+		{
+			name:      "set new key with different existing key",
+			initialMD: metadata.MD{"key2": []string{"value2"}},
+			setKey:    "key1",
+			setValue:  "value1",
+			wantValue: "value1",
+			wantKeys:  []string{"key2", "key1"},
+		},
+		{
+			name:      "set non-grcp-trace-bin binary key",
+			initialMD: metadata.MD{"key1": []string{"value1"}},
+			setKey:    "non-grcp-trace-bin",
+			setValue:  base64.StdEncoding.EncodeToString([]byte{0x01, 0x02, 0x03}),
+			wantValue: "", // non-grcp-trace-bin key should not be set
+			wantKeys:  []string{"key1"},
```

Comment:
Same here:
* grpc-trace-bin
* more than one key/value pair that overwrites
* Both grpc-trace-bin and another string key

Reply:
yeah moved grpc-trace-bin under this test and deleted the separate test. Again, the difference is that if test case key -bin, we need to verify the outgoing trace header from stats package and return. For other cases, we verify the context's metadata

Reply:
> more than one key/value pair that overwrites

override existing key is the case which overwrite an existing key with a new value

Reply:
> Both grpc-trace-bin and another string key

didn't quite get you what you mean here. The -bin test cases already has initial md with string key value pair and we verify if both keys are present

Reply:
What I mean is if you get grpc-trace-bin and another string key, what is the expected output?

Reply:
yeah the test case "grpc-trace-bin binary key" already has initialMD with a string key so if grpc-trace-bin comes, it will have both key1 and grpc-trace-bin in want keys and want value will be expected value for grpc-trace-bin